### ISSUE TITLE
Ways to retain request socket data received in preread_by_lua* for later phases?

### ISSUE THREAD

**Issue by klzgrad** [openresty/stream-lua-nginx-module#80]
The use case is essentially the same as `ngx_stream_ssl_preread_module` - inspect the initial data and `proxy_pass` everything received intact to the backend derived in `balancer_by_lua*`.

So far the only way to inspect data in preread_by_lua is through `ngx.req.socket`. The doc says "calling the receive() method of the request socket will consume the data from the buffer and those data will not be seen by handlers further down the chain." The part that is inspected for balancing is unable to be relayed during `proxy_pass`.

I would like to be able to `proxy_pass` everything. How do I do that?

**Comment by dndx** [comment]
Unfortunately the current API does not support this. See:

https://github.com/openresty/stream-lua-nginx-module#preread_by_lua_block

> However, keep in mind that calling the receive() method of the request socket will consume the data from the buffer and those data will not be seen by handlers further down the chain.

There is currently a plan to improve this by implementing `reqsock:peek()` method that will read the data but not consume it from the preread buffer. It's not a super high priority but definitely on the list.

If you are able to contribute a patch, that would also be highly appreciated.

**Comment by klzgrad** [comment]
Can you elaborate on this design?

I'm asking because right now reqsock:receive() uses its own buffer if I read it correctly. No data is put in the preread buffer (`session->connection->buffer`) at all by lua handlers (though ngx_stream_ssl_preread_module totally uses the preread buffer). An example: `reqsock:receive(1)` during preread will make more than 1 byte unavailable for later phases, because `recv()` puts more than 1 byte of data into the lua socket buffer and that data is not passed to the preread buffer.

A design of a peek method like this would imply certain kind of synchronization between the two buffers. I'm afraid that'd be too error prone for me to implement. But anyway, right now I'm using a hack to directly push data to the preread buffer from lua.

**Comment by devdavidkarlsson** [comment]
@dndx Where can one look at the :receive() method, we also really want the peek function. F eks. to perform authentication of streams in the nginx... :

**Comment by dndx** [comment]
@devdavidkarlsson Hello,

You can take a look at https://github.com/openresty/lua-nginx-module#ngxreqsocket for usage of the `:receive` method.

**Comment by devdavidkarlsson** [comment]
Ok, sorry for some newbie questions here so the: "peek" method would need to be included in the lua-nginx-module, and not just in stream-lua-nginx-module?

Still having some trouble navigating the source code, in the lua-nginx-module: 
https://github.com/openresty/lua-nginx-module/blob/0b514d3951b5597c67e678134c01da870ea88a17/src/ngx_http_lua_socket_tcp.c
there is a method for **ngx_http_lua_socket_tcp_read**

Is that what is being used by streams also?

**Comment by agentzh** [comment]
@devdavidkarlsson Both of the http and stream lua modules share the same Lua API (mostly). That's why @dndx gave you a link to the http lua module's API docs for the stream lua module's same name Lua API.

**Comment by agentzh** [comment]
@devdavidkarlsson It makes no sense for you to look into http lua's implementation for stream lua's same name Lua API.

**Comment by dndx** [comment]
@devdavidkarlsson No, `:peek()` will not be implemented for `lua-nginx-module`. In fact, it should only be available for stream request socket inside the `preread` phase.

**Comment by yigangbupt** [comment]
> The use case is essentially the same as `ngx_stream_ssl_preread_module` - inspect the initial data and `proxy_pass` everything received intact to the backend derived in `balancer_by_lua*`.
> 
> So far the only way to inspect data in preread_by_lua is through `ngx.req.socket`. The doc says "calling the receive() method of the request socket will consume the data from the buffer and those data will not be seen by handlers further down the chain." The part that is inspected for balancing is unable to be relayed during `proxy_pass`.
> 
> I would like to be able to `proxy_pass` everything. How do I do that?


Have you solved the problem?

**Comment by yigangbupt** [comment]
> Fixed in [7637499](https://github.com/openresty/stream-lua-nginx-module/commit/7637499991e9885a75e69ba4e763feac1a56c448).

sorry, when i use receive() in "preread", proxy_pass is failed. The method peek() is not implemented yet. Is there any other way to solve this problem？

**Comment by dndx** [comment]
@yigangbupt `peek()` has just been merged to master and will be released in the next OpenResty release. Or you can compile from master and start using it today.

**Comment by jans1086** [comment]
how can i  get  all of  the  http headers  with peek(), when i use stream
